### ISSUE TITLE
/v1/catalog/nodes sometimes returns 0 nodes if there's a leadership change in Consul

### ISSUE THREAD

**Issue by sgirones** [hashicorp/consul#4103]
If you have a question, please direct it to the
[consul mailing list](https://www.consul.io/community.html) if it hasn't been
addressed in either the [FAQ](https://www.consul.io/docs/faq.html) or in one
of the [Consul Guides](https://www.consul.io/docs/guides/index.html).

When filing a bug, please include the following:

### Description of the Issue (and unexpected/desired result)
When using consul-template + consul, some queries to the Catalog API return less than expected or even 0 nodes (expected is 132). We detected this during a leadership change in the cluster and then we have been able to reproduce it by randomly killing Consul servers

There's an open issue in the consul-template repo but wanted to open it here as well as it may be a Consul issue instead: https://github.com/hashicorp/consul-template/issues/1098

### Reproduction steps
Randomly kill the Consul servers while you have a consul-template watching the Catalog API (more details in the contul-template repo). We currently do it by running the following command in our 3 servers:
```
while true; do sudo pkill consul$; sleep $(( $RANDOM % 8 + 2 )); done
```

We cannot reproduce it consistently.

### `consul version` for both Client and Server
Client: `Consul v1.0.6`
Server: `Consul v1.0.6`

### `consul info` for both Client and Server
Client:
```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 2
	services = 2
build:
	prerelease =
	revision = 9a494b5f
	version = 1.0.6
consul:
	known_servers = 3
	server = false
runtime:
	arch = amd64
	cpu_count = 4
	goroutines = 45
	max_procs = 4
	os = linux
	version = go1.9.3
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 3768
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 13981
	members = 131
	query_queue = 0
	query_time = 29
```

Server:
```
agent:
	check_monitors = 0
	check_ttls = 1
	checks = 1
	services = 1
build:
	prerelease =
	revision = 9a494b5f
	version = 1.0.6
consul:
	bootstrap = false
	known_datacenters = 1
	leader = true
	leader_addr = ip-3:8300
	server = true
raft:
	applied_index = 14684725
	commit_index = 14684725
	fsm_pending = 0
	last_contact = 0
	last_log_index = 14684725
	last_log_term = 3966
	last_snapshot_index = 14684631
	last_snapshot_term = 3966
	latest_configuration = [{Suffrage:Voter ID:539c6dc4-6cf5-1282-5091-379dd1d89dd1 Address:ip-1:8300} {Suffrage:Voter ID:09bb9e85-b072-b744-f6cc-1fc67ea7f660 Address:ip-2:8300} {Suffrage:Voter ID:facb3bb6-5174-1e1d-184e-486d68325da1 Address:ip-3:8300}]
	latest_configuration_index = 13596262
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 3966
runtime:
	arch = amd64
	cpu_count = 2
	goroutines = 797
	max_procs = 2
	os = linux
	version = go1.9.3
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 3768
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 13981
	members = 131
	query_queue = 0
	query_time = 29
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 7988
	members = 3
	query_queue = 0
	query_time = 1
```

### Log Fragments or Link to [gist](https://gist.github.com/)

Consul-template logs: https://gist.github.com/sgirones/74eb8088e7a46833a9a3ed75295754f3

**Comment by sgirones** [comment]
We experienced the same issue while watching nodes and services, but not a KV key. This section in the FAQ ([link](https://www.consul.io/docs/faq.html#q-is-consul-eventually-or-strongly-consistent-)) claims that the Catalog is built using the Gossip protocol.
Is there a known scenario where the Catalog becomes empty or is rebuilt from scratch?

**Comment by sgirones** [comment]
I see these errors around the same time, although not always:
```
[ERR] consul.acl: Failed to get policy from ACL datacenter: No cluster leader
[DEBUG] consul: dropping node "node-1" from result due to ACLs
[DEBUG] consul: dropping node "node-2" from result due to ACLs
[...]
```

Looking at the code, that happens inside the `filterMembers` method:
https://github.com/hashicorp/consul/blob/master/agent/agent_endpoint.go#L238
which is called inside the endpoint `/v1/agent/members`.

Is it possible that the catalog is rebuilt using that endpoint at some point and given it returns 0 nodes, the catalog is emptied? Someone with better knowledge of the internal may be able to help here.

**Comment by sgirones** [comment]
After digging more into the code, what is happening is that the API call result is filtered using an ACL token, so if the cluster has no leader, you will receive a list of 0 nodes depending on the `acl_down_policy` config:

https://github.com/hashicorp/consul/blob/master/agent/consul/catalog_endpoint.go#L202

In our case we had `extend-cache`, so I assume the cluster had no leader and the server replying to the API call still couldn't cache any ACL token.

This is unfortunate for service discovery purposes, where you need a reliable source of truth but at the same time you don't want to allow every operation on the cluster when there's no leader.

Would it make sense to have a ReadOnly policy for `acl_down_policy`?

**Comment by sgirones** [comment]
Closed in favour of #4139
